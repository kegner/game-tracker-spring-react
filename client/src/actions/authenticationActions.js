export const SET_USER = "SET_USER";
export const CLEAR_USER = "CLEAR_USER";

export const setUserData = (user) => ({
  type: SET_USER,
  payload: { user }
});

export const clearUserData = () => ({
    type: CLEAR_USER
});