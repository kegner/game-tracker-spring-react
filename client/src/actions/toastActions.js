export const SHOW_SUCCESS = "SHOW_SUCCESS";
export const SHOW_ERROR = "SHOW_ERROR";
export const HIDE_TOAST = "HIDE_TOAST";

export const showSuccessToast = (message) => ({
  type: SHOW_SUCCESS,
  payload: { message }
});

export const showErrorToast = (message) => ({
  type: SHOW_ERROR,
  payload: { message }
});

export const hideToast = (id) => ({
  type: HIDE_TOAST,
  payload: { id }
});