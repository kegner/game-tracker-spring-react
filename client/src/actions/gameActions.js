import axios from 'axios';
import { showSuccessToast, showErrorToast } from './toastActions';

export const SELECT_GAME = "SELECT_GAME";
export const FETCH_GAMES = "FETCH_GAMES";
export const CREATE_GAME = "CREATE_GAME";
export const UPDATE_GAME = "UPDATE_GAME";
export const DELETE_GAME = "DELETE_GAME";
export const START_LOADING = "START_LOADING";
export const START_SAVING = "START_SAVING";
export const STOP_LOADING = "STOP_LOADING";
export const STOP_SAVING = "STOP_SAVING";

const fetchSuccess = (games) => ({
  type: FETCH_GAMES,
  payload: { games }
});

const createSuccess = (game) => ({
  type: CREATE_GAME,
  payload: { game }
});

const updateSuccess = (game) => ({
  type: UPDATE_GAME,
  payload: { game }
});

const deleteSuccess = (id) => ({
  type: DELETE_GAME,
  payload: { id }
});

const startLoading = () => ({
  type: START_LOADING
});

const startSaving = () => ({
  type: START_SAVING
});

const stopLoading = () => ({
  type: STOP_LOADING
});

const stopSaving = () => ({
  type: STOP_SAVING
});

const errorHandler = (error, dispatch) => {
  const validResponses = [404, 409];
  if (error && error.response && validResponses.includes(error.response.status)) {
    // this is the expected error format for Spring exception messages
    dispatch(showErrorToast(error.response.data.message));
  } else {
    console.error(error);
  }
};

export const selectGame = (game) => ({
  type: SELECT_GAME,
  payload: { game }
});

export const fetchGames = (userId) => async (dispatch) => {
  dispatch(startLoading());

  try {
    const res = await axios.get("/api/v1/games", { params: { userId } });
    dispatch(fetchSuccess(res.data.games));
  } catch (error) {
    dispatch(stopLoading());
    errorHandler(error, dispatch);
  }
};

export const createGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.post("/api/v1/games", game);

    dispatch(createSuccess(res.data));
    dispatch(showSuccessToast("Game added."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const updateGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.put("/api/v1/games", game);

    dispatch(updateSuccess(res.data));
    dispatch(showSuccessToast("Game updated."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const deleteGame = (id) => async (dispatch) => {
  dispatch(startSaving());

  try {
    await axios.delete(`/api/v1/games/${id}`);
    dispatch(deleteSuccess(id));
    dispatch(showSuccessToast("Game deleted."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};