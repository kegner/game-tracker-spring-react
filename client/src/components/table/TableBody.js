import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { selectGame } from '../../actions/gameActions';
import DeleteButton from './DeleteButton';

const TableBody = (props) => {

  const { gamesPages, pageIndex } = props;

  const { selectedGame } = useSelector(state => state.gameReducer);

  const dispatch = useDispatch();

  const rowClickHandler = (game) => {
    dispatch(selectGame(game));
  }

  return (
    <tbody>
      {gamesPages[pageIndex].map(game => {
        const activeClass = selectedGame.id === game.id ? "table-active" : "";
        return <tr key={game.id} className={activeClass} onClick={() => rowClickHandler(game)}>
          <td className="text-center">{game.title}</td>
          <td className="text-center">{game.status}</td>
          <td className="text-center">{game.rating}</td>
          <td className="text-center"><DeleteButton id={game.id} /></td>
        </tr>
      })}
    </tbody>
  )
}

export default TableBody
