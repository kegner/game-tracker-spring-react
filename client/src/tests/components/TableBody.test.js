import React from 'react';
import { Provider } from "react-redux";
import { render, fireEvent } from '@testing-library/react';
import { SELECT_GAME } from '../../actions/gameActions';
import { createTestStore } from '../store/testStore';
import TableBody from '../../components/table/TableBody';

let store;
let dispatch;
let component;
let container;
let asFragment;
let game = {
  id: "123", title: "My Game"
};
let gamesPages = [[game]];

describe("TableBody", () => {

  beforeEach(() => {
    store = createTestStore({
      mockStore: true,
      initialState: {
        gameReducer: {
          selectedGame: game
        }
      }
    });
    dispatch = store.dispatch;
    component = render(
      <Provider store={store}>
        <TableBody gamesPages={gamesPages} pageIndex={0} />
      </Provider>
    );
    container = component.container;
    asFragment = component.asFragment;
  });

  it("exists", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("has the correct row count", () => {
    const rowCount = container.querySelectorAll("tr").length;
    expect(rowCount).toBe(1);
  });

  it("selects a row", async () => {
    const expectedActions = [
      { type: SELECT_GAME, payload: { game }},
    ];

    const rowElement = container.querySelector("tr");
    expect(dispatch).not.toHaveBeenCalled();
    fireEvent.click(rowElement);
    // await for all dispatches
    await new Promise(r => setTimeout(r, 1));
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toEqual(expectedActions);
  });

  it("selects a new row and sets active class", async () => {
    store = createTestStore();
    dispatch = store.dispatch;
    component = render(
      <Provider store={store}>
        <TableBody gamesPages={gamesPages} pageIndex={0} />
      </Provider>
    );
    container = component.container;
    asFragment = component.asFragment;

    const rowElement = container.querySelector("tr");
    let selectedRowCount = container.querySelectorAll(".table-active").length;
    expect(selectedRowCount).toBe(0);
    fireEvent.click(rowElement);
    // await for all dispatches
    await new Promise(r => setTimeout(r, 1));
    selectedRowCount = container.querySelectorAll(".table-active").length;
    expect(selectedRowCount).toBe(1);
  });

});