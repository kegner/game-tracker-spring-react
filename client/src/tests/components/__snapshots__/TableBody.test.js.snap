// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBody exists 1`] = `
<DocumentFragment>
  My Game
  <button
    aria-label="Delete"
    class="btn btn-danger"
  >
    <i
      class="bi bi-trash"
    />
  </button>
</DocumentFragment>
`;
